### PR TITLE
Merge param descriptions when adding or updating types

### LINKED ISSUE #9440: Psalter: Docblock param descriptions duplicated or removed
I found two issues when getting Psalter to add/fix parameters into doc blocks (either `MissingParamType` or `MismatchingDocblockParamType`, both with `safe_types`) where the existing param descriptions are either ignored or removed.

I made one issue as these seem related. Happy to split if you prefer.

### Issue 1: Param is duplicated

If I have the following code and I run Psalter to add a type in then it duplicates the param in the doc block rather than merging it.

```php
class C {
    /**
     * @param $ab the string you pass in
     */
    public function fooFoo($ab): void {}
}

/**
 * @param string $ab
 */
function callsWithString($ab): void {
    (new C)->fooFoo($ab);
}
```
is turned into
```php
class C {
    /**
     * @param $ab the string you pass in
     * @param string $ab
     */
    public function fooFoo($ab): void {}
}

/**
 * @param string $ab
 */
function callsWithString($ab): void {
    (new C)->fooFoo($ab);
}
```

Where as I would hope would be that the docblock would become
```php
/**
 * @param string $ab the string you pass in
 */
```

### Issue 2: Param description removed when fixing incorrect doc block types

Worse, and somewhat relatedly, if Psalter is fixing a return type, it drops the param description entirely.

```php
/**
 * @param int $s the string
 */
function foo(string $s): string {
    return "hello";
}
```
becomes
```php
/**
 * @param string $s
 */
function foo(string $s): string {
    return "hello";
}
```

### PR DESCRIPTION
**Pull request by aboyton**
Fixes #9440

### CODE REVIEW

**Comment by orklah** [comment]
Seems great! I'm surprised about the error on CircleCI, I think this was fixed some times ago. Can you rebase on up to date master to see if it gets fixed?

File: src/Psalm/Internal/FileManipulation/FunctionDocblockManipulator.php [hunk 1]
```diff
@@ -332,7 +334,29 @@ private function getDocblock(): string
...
+                            $modified_docblock = true;
+                        }
+
+                        $param_block = $new_param_block;
+                        $found_in_params = true;
+                        break;
+                    }
+
+                    // If there is a type
                     if (($doc_parts[1] ?? null) === '$' . $param_name) {
```
Comment (weirdan) [inline src/Psalm/Internal/FileManipulation/FunctionDocblockManipulator.php:354]: Doesn't it actually handle the case when there's *no* type?

Comment (weirdan) [reply src/Psalm/Internal/FileManipulation/FunctionDocblockManipulator.php:354]: Ah, ignore me. `$doc_parts[0]` is (normally) a type, and `$doc_parts[1]` is the param name.

**Comment by aboyton** [comment]
> Seems great! I'm surprised about the error on CircleCI, I think this was fixed some times ago. Can you rebase on up to date master to see if it gets fixed?

I force pushed an update, not sure if that's correct. Turns out I'm accidentally still on the branch I made from my last PR. Let me know if this hasn't worked properly.

**Comment by weirdan** [comment]
> I force pushed an update, not sure if that's correct

It's fine.

> Turns out I'm accidentally still on the branch I made from my last PR

Not a problem either.

All is good now.

**Review by weirdan — APPROVED** [review]

**Comment by aboyton** [comment]
Thanks for the quick merge. How often do you cut a new release?

**Comment by orklah** [comment]
Note: https://github.com/vimeo/psalm/issues/4192 should be fixed or almost fixed:
@aboyton if you still feel like working on it, you could try to check if this snippet https://psalm.dev/r/1d19c39522 with --safe-types=true fixes the partial @param annotation correctly :)

**Comment by aboyton** [comment]
I somehow completely missed #4192. Yes, the code from https://psalm.dev/r/1d19c39522 with `--safe-types=true` used to return

```php
/**
 * @param $x
 * @param bool $x
 */
function foo($x) {}
```
and with this change it now returns
```php
/**
 * @param bool $x
 */
function foo($x) {}
```
as expected, so I think #4192 can be closed now too.

**Comment by orklah** [comment]
Cool! Thanks!

**Comment by weirdan** [comment]
> How often do you cut a new release?

There's no fixed schedule, as @orklah said, but lately it's been once in 2-4 weeks unless some critical issue appears.

**Pull request merged**